ecosystem/indexer-grpc/indexer-grpc-manager/src/metadata_manager.rs

```diff
@@ -0,0 +1,383 @@
+            master_address: None,
+        };
+        let service_info = ServiceInfo {
+            address: Some(self.self_advertised_address.clone()),
+            info: Some(Info::GrpcManagerInfo(grpc_manager_info)),
+        };
+        let request = HeartbeatRequest {
+            service_info: Some(service_info),
+        };
+        let _ = client.heartbeat(request).await?;
+
+        Ok(())
+    }
+
+    async fn ping_fullnode(
+        &self,
+        address: String,
+        mut client: FullnodeDataClient<Channel>,
+    ) -> Result<()> {
+        trace!("Pinging fullnode {address}.");
+        let request = PingFullnodeRequest {};
+        let response = client.ping(request).await?;
+        if let Some(info) = response.into_inner().info {
+            self.handle_fullnode_info(address, info)
+        } else {
+            bail!("Bad response.")
+        }
+    }
+
+    async fn ping_live_data_service(
+        &self,
+        address: String,
+        mut client: DataServiceClient<Channel>,
+    ) -> Result<()> {
+        let request = PingDataServiceRequest {
+            known_latest_version: Some(self.get_known_latest_version()),
+            ping_live_data_service: true,
+        };
+        let response = client.ping(request).await?;
+        if let Some(info) = response.into_inner().info {
+            match info {
+                aptos_protos::indexer::v1::ping_data_service_response::Info::LiveDataServiceInfo(info) => {
+                    self.handle_live_data_service_info(address, info)
+                },
+                _ => bail!("Bad response."),
+            }
+        } else {
+            bail!("Bad response.")
+        }
+    }
+
+    async fn ping_historical_data_service(
+        &self,
+        address: String,
+        mut client: DataServiceClient<Channel>,
+    ) -> Result<()> {
+        let request = PingDataServiceRequest {
+            known_latest_version: Some(self.get_known_latest_version()),
+            ping_live_data_service: false,
+        };
+        let response = client.ping(request).await?;
+        if let Some(info) = response.into_inner().info {
+            match info {
+                aptos_protos::indexer::v1::ping_data_service_response::Info::HistoricalDataServiceInfo(info) => {
+                    self.handle_historical_data_service_info(address, info)
+                },
+                _ => bail!("Bad response."),
+            }
+        } else {
+            bail!("Bad response.")
+        }
+    }
+
+    fn handle_live_data_service_info(
+        &self,
+        address: String,
+        info: LiveDataServiceInfo,
+    ) -> Result<()> {
+        let mut entry = self
+            .live_data_services
+            .entry(address.clone())
+            .or_insert(LiveDataService::new(address));
+        entry.value_mut().recent_states.push_back(info);
+        if entry.value().recent_states.len() > MAX_NUM_OF_STATES_TO_KEEP {
```

Comment:
Why do we need to keep more than 1 state history?

Reply:
A sample use case is I can calculate the tps. (especially for debugging purpose, this can have much more information and granularity than grafana metric)